pkg/auth/handler/sso/auth_handler.go

```diff
@@ -0,0 +1,456 @@
+			return resp, err
+		}
+
+		if err != skydb.ErrUserNotFound {
+			// some other error
+			return resp, err
+		}
+
+		// check if user is already linked
+		userID := oauthAuthInfo.State.UserID // skygear userID
+		_, err = h.OAuthAuthProvider.GetPrincipalByUserID(userID)
+		if err == nil {
+			err = skyerr.NewError(skyerr.InvalidArgument, "provider account already linked with existing user")
+			return resp, err
+		}
+
+		if err != skydb.ErrUserNotFound {
+			// some other error
+			return resp, err
+		}
+
+		var info authinfo.AuthInfo
+		if err = h.AuthInfoStore.GetAuth(userID, &info); err != nil {
+			err = skyerr.NewError(skyerr.ResourceNotFound, "user not found")
+			return resp, err
+		}
+
+		_, err = h.createPrincipalByOAuthInfo(info.ID, oauthAuthInfo)
+		if err != nil {
+			return resp, err
+		}
+		resp = map[string]interface{}{
+			"result": "OK",
+		}
+	}
+
+	return
+}
+
+func (h AuthHandler) handleLogin(info *authinfo.AuthInfo, oauthAuthInfo sso.AuthInfo) (err error) {
+	now := timeNow()
+
+	principal, err := h.OAuthAuthProvider.GetPrincipalByProviderUserID(oauthAuthInfo.ProviderName, oauthAuthInfo.ProviderUserID)
+	if err != nil {
+		if err != skydb.ErrUserNotFound {
+			return
+		}
+		err = nil
+	}
+
+	if principal == nil {
+		principal, err = h.authLinkUser(oauthAuthInfo)
+		if err != nil {
+			return
+		}
+	}
+
+	if principal == nil {
+		// if there is no existed user
+		// signup a new user
+		*info = authinfo.NewAuthInfo()
+		info.LastLoginAt = &now
+
+		// Get default roles
+		defaultRoles, e := h.RoleStore.GetDefaultRoles()
+		if e != nil {
+			err = skyerr.NewError(skyerr.InternalQueryInvalid, "unable to query default roles")
+			return
+		}
+
+		// Assign default roles
+		info.Roles = defaultRoles
+
+		// Create AuthInfo
+		if e = h.AuthInfoStore.CreateAuth(info); e != nil {
+			if e == skydb.ErrUserDuplicated {
+				err = skyerr.NewError(skyerr.Duplicated, "user duplicated")
+				return
+			}
+			// TODO:
+			// return proper error
+			err = skyerr.NewError(skyerr.UnexpectedError, "Unable to save auth info")
+			return
+		}
+
+		_, err = h.createPrincipalByOAuthInfo(info.ID, oauthAuthInfo)
+	} else {
+		principal.AccessTokenResp = oauthAuthInfo.ProviderAccessTokenResp
+		principal.UserProfile = oauthAuthInfo.ProviderUserProfile
+		principal.UpdatedAt = &now
+
+		if err = h.OAuthAuthProvider.UpdatePrincipal(principal); err != nil {
+			err = skyerr.MakeError(err)
+			return
+		}
+
+		if e := h.AuthInfoStore.GetAuth(principal.UserID, info); e != nil {
+			if err == skydb.ErrUserNotFound {
+				err = skyerr.NewError(skyerr.ResourceNotFound, "User not found")
+				return
+			}
+			err = skyerr.NewError(skyerr.ResourceNotFound, "User not found")
+			return
+		}
+	}
+	return
+}
+
+func (h AuthHandler) authLinkUser(oauthAuthInfo sso.AuthInfo) (*oauth.Principal, error) {
+	principals, e := h.PasswordAuthProvider.GetPrincipalsByAuthData(oauthAuthInfo.ProviderAuthData)
```

Comment:
In the original SSO plugin, auto link user is done by the configuration `SSO_AUTO_LINK_PROVIDER_KEYS` dynamically. User can disable this by providing empty config.

Also, the case becomes more complicated in the new password provider models. 
    1. this function should not return multiple principals
    2. we should create password provider entry with empty password to occupy the email... etc

I think it is a bit too much to handle it here in this PR? Maybe we should remove the auto link feature from this PR and handle it in https://github.com/SkygearIO/skygear-server/issues/837